### PR TITLE
Fix `generateCatalogIndex` (post search provider changes)

### PR DESCRIPTION
**Pull request by nf-s**
### Fix `generateCatalogIndex` (post search provider changes)

I have also added `commander` to parse args more nicely - this is what the `-h` help looks like

```
Usage: generateCatalogIndex [options]

Generate catalog index (**experimental**)

This will "crawl" a terria JS catalog, load all groups and references and then create an "index" file which contains fully resolved tree of models.

Example usage
- node ./build/generateCatalogIndex.js -c http://localhost:3001/config.json -b http://localhost:3001/
  

Options:
  -c, --configUrl <configUrl>  configUrl URL to map-config
  -b, --baseUrl <baseUrl>      baseUrl will be used as:
                               - `origin` property for CORS
                               - URL for `serverConfig`
                               - URL for `proxy`
  -o, --outPath [outPath]      catalog-index JSON file path (default:
                               "catalog-index.json")
  --basicAuth [basicAuth]      basic auth token to add to requests which include
                               `baseUrl` (or `proxy/`)
  --excludeIds [ids...]        CSV of model IDs to exclude from catalog index (eg
                               "some-id-1 some-id-2")
  -s, --speed [speed]          speed will control number of concurrently loaded catalog
                               groups/references:
                               - default value is 1 (which is around 10 loads per
                               second)
                               - minimum value is 1
                               - If speed = 10 - then expect around 100 loads per second
                               - Note: loads are somewhat randomised across catalog, so
                               you don't hit one server with many requests
                               - Also note: one load may not equal one request. some
                               groups/references do not make network requests (default:
                               1)
  -t, --timeout [timeout]      Network request timeout (in ms) (default: 30000)
  -h, --help                   display help for command
```

If you want to test it

- Checkout this branch
- cd into `terriajs`
- `yarn run build-tools`
- `node ./build/generateCatalogIndex.js -c http://localhost:3001/config.json -b http://localhost:3001/`

Alternatively - look at this github workflow using this branch

- https://github.com/TerriaJS/saas-catalogs-public/actions/runs/8368761624

### Checklist

- [ ] ~There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)~
- [x] I've updated relevant documentation in `doc/`.
- [x] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: buildprocess/configureWebpack.js [hunk 1]
```diff
@@ -160,6 +162,7 @@ function configureWebpack(
   config.module.rules.push({
     test: /\.(ts|js)x?$/,
     include: [
+      path.resolve(terriaJSBasePath, "node_modules", "commander"),
```
Comment (nf-s) [inline buildprocess/configureWebpack.js:165]: These changes are needed to fix a webpack error - there is probably a better way to handle the `commander` path here

```
ERROR in ./node_modules/commander/lib/command.js 380:47
Module parse failed: Unexpected token (380:47)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|     }
| 
>     enableOrNameAndArgs = enableOrNameAndArgs ?? 'help [command]';
|     const [, helpName, helpArgs] = enableOrNameAndArgs.match(/([^ ]+) *(.*)/);
|     const helpDescription = description ?? 'display help for command';
 @ ./node_modules/commander/index.js 2:20-47
 @ ./buildprocess/generateCatalogIndex.ts
```

**Review by mwu2018 — CHANGES_REQUESTED** [review]
Need to update `.npmrc` to specify node version 18.

**Comment by mwu2018** [comment]
When running `node ./build/generateCatalogIndex.js -c http://localhost:3001/config.json -b http://localhost:3001/`, got the following content in file `catalog-indexerrors.json` but the group and its members are all working.

```
Errors
Nested error:
  FAILED to load GROUP Marine water quality grids for the Great Barrier Reef region - monthly data (///FxPHcHvxoO)
    Failed to load group `Marine water quality grids for the Great Barrier Reef region - monthly data`
      Failed to load `Marine water quality grids for the Great Barrier Reef region - monthly data` metadata
        core.terriaError.networkRequestMessage
          Request has failed.
```

@nf-s Is this error expected? If it can be ignored, the changes are good to be merged.

**Comment by nf-s** [comment]
Good find @mwu2018 !
There was a bug where relative URLs were failing - I have added a fix to `patchNetworkRequests` to convert them to absolute URLs

**Comment by mwu2018** [comment]
With the fixes, the above errors have gone but new error occurred.
```
Errors
Nested error:
  FAILED to load GROUP Sand Content [60 - 100 cm] (///5z6OTFssaW/5z6OTFssaW/SND_ACLEP_AU_SAT_D/MapServer/5z6OTFssaW/SND_ACLEP_AU_SAT_D/MapServer/16)
    Failed to load group `Sand Content [60 - 100 cm]`
      Failed to load `Sand Content [60 - 100 cm]` metadata
        core.terriaError.networkRequestMessage
          Request has failed. Status Code: 502
  FAILED to load GROUP Effective Cation Exchange Capacity [60 - 100 cm] (///5z6OTFssaW/5z6OTFssaW/ECE_ACLEP_AU_TRN_N/MapServer/5z6OTFssaW/ECE_ACLEP_AU_TRN_N/MapServer/16)
    Failed to load group `Effective Cation Exchange Capacity [60 - 100 cm]`
      Failed to load `Effective Cation Exchange Capacity [60 - 100 cm]` metadata
        core.terriaError.networkRequestMessage
          Request has failed. Status Code: 502
  FAILED to load GROUP Available Phosphorus [0 - 5 cm] (///5z6OTFssaW/5z6OTFssaW/AVP_ACLEP_AU_TRN_N/MapServer/5z6OTFssaW/AVP_ACLEP_AU_TRN_N/MapServer/0)
    Failed to load group `Available Phosphorus [0 - 5 cm]`
      Failed to load `Available Phosphorus [0 - 5 cm]` metadata
        core.terriaError.networkRequestMessage
          Request has failed. Status Code: 502
```

It appears that the indexer did not handle deeply nested layers well.

**Comment by nf-s** [comment]
If requests are returning 502, then it is a server-side issue

**Comment by mwu2018** [comment]
I can see those layers work well. E.g., This layer is a subgroup: https://www.asris.csiro.au/arcgis/rest/services/TERN/SND_ACLEP_AU_SAT_D/MapServer/16

**Comment by nf-s** [comment]
I re-ran the indexer and got no-errors - must have been intermittent server issue

```
node ./build/generateCatalogIndex.js -c http://localhost:3001/config.json -b http://localhost:3001/ --timeout 60000
```

**Comment by mwu2018** [comment]
Looks like intermittent issue because other layers have a very similar structure but have no issues.

**Review by mwu2018 — APPROVED** [review]
LGTM after running local test.

**Pull request merged**